src/window/building/distribution.c

```diff
@@ -1124,3 +1143,112 @@ void window_building_draw_mess_hall(building_info_context *c)
+                0, "Areldir", "Econ_Logistics", "Trade Policy 1", 0
+        },
+        {
+                0, TR_BUILDING_CARAVANSERAI_POLICY_2_TITLE, TR_BUILDING_CARAVANSERAI_POLICY_2,
+                0, "Areldir", "Econ_Logistics", "Trade Policy 2", 0
+        },
+        {
+                0, TR_BUILDING_CARAVANSERAI_POLICY_3_TITLE, TR_BUILDING_CARAVANSERAI_POLICY_3,
+                0, "Areldir", "Econ_Logistics", "Trade Policy 3", 0
+        }
+};
+
+static void draw_policy_image_border(int x, int y, int focused)
+{
+    int id = assets_get_image_id(assets_get_group_id("Areldir", "Econ_Logistics"),
+                        "Policy Selection Borders");
+
+    image_draw(id + focused, x, y);
+    image_draw(id + 2 + focused, x + 105, y + 5);
+    image_draw(id + 4 + focused, x, y + 90);
+    image_draw(id + 6 + focused, x, y + 5);
+}
+
+void window_building_draw_caravanserai_action(building_info_context *c)
+{
+
+    int policy = building_monument_module_type(BUILDING_CARAVANSERAI);
+    int policy_title = TR_BUILDING_CARAVANSERAI_NO_POLICY;
+
+    if (policy) {
+        policy_title = caravanserai_policy_options[policy-1].header;
+    }
+
+    text_draw_multiline(translation_for(policy_title), c->x_offset + 160, c->y_offset + 160,
+                       260, FONT_NORMAL_BLACK, 0);
+
+    if (policy) {
+        text_draw_multiline(translation_for(policy_title + 1), c->x_offset + 160, c->y_offset + 185,
+                            260, FONT_NORMAL_BLACK, 0);
+
+        int policy_image = assets_get_image_id(assets_get_group_id((char *) caravanserai_policy_options[policy-1].asset_author, (char *) caravanserai_policy_options[policy-1].asset_name),
+                                               (char *) caravanserai_policy_options[policy-1].asset_image_id);
+
+        image_draw(policy_image, c->x_offset + 32, c->y_offset + 150);
+    } else {
+        int policy_image = assets_get_image_id(assets_get_group_id("Areldir", "Econ_Logistics"), "Trade Policy");
+        image_draw(policy_image, c->x_offset + 32, c->y_offset + 150);
+    }
+
+}
+
+void window_building_draw_caravanserai_foreground(building_info_context *c)
+{
+    draw_policy_image_border(c->x_offset + 32, c->y_offset + 150, data.caravanserai_focus_button_id == 1);
+}
+
+static void caravanserai_policy(int selection)
+{
+    if (!selection) {
+        return;
+    }
+
+    sound_speech_play_file("wavs/market4.wav");
+    building *b = building_get(city_buildings_get_caravanserai());
+
+    city_finance_process_construction(TRADE_POLICY_COST);
+    building_monument_add_module(b, selection);
+}
+
+static void caravanserai_action(int param1, int param2)
+{
+    option_menu_item options[3] = { caravanserai_policy_options[0], caravanserai_policy_options[1], caravanserai_policy_options[2] };
+    window_action_popup_show(TR_BUILDING_CARAVANSERAI_POLICY_TITLE, TR_BUILDING_CARAVANSERAI_POLICY_TEXT, options, caravanserai_policy, 1);
```

Comment:
Why not directly send `caravanserai_policy_options` as a parameter?

Reply:
I'm writing like in culture.c > add_module_prompt

Reply:
Ah I didn't check that one, probably the original code issue is from there. Still, better fix one than none!